File: Coding_Problems/2_maxSumSubArray.js
function maxSumSubarray(arr, k) {
  let windowSum = 0;
  let maxSum = 0;

  // Build the first window
  for (let i = 0; i < k; i++) {
    windowSum += arr[i];
  }

  maxSum = windowSum;

  // Slide the window
  for (let i = k; i < arr.length; i++) {
    windowSum += arr[i]; // add new element on right
    windowSum -= arr[i - k]; // remove old element on left
    maxSum = Math.max(maxSum, windowSum);
  }

  return maxSum;
}

let arr = [1, 2, 3, 4, 5];
let k = 2;
console.log(maxSumSubarray(arr, k)); // Output: 9 (4 + 5)

File: Coding_Problems/2_maxSumSubArray.cpp
#include <iostream>
#include <vector>

using namespace std;

int maxSumSubarray(vector<int>& arr, int k) {
    int windowSum = 0;
    int maxSum = 0;

    // Build first window
    for (int i = 0; i < k; i++) {
        windowSum += arr[i];
    }

    maxSum = windowSum;

    // Slide the window
    for (int i = k; i < arr.size(); i++) {
        windowSum += arr[i];
        windowSum -= arr[i - k];

        maxSum = maxSum>windowSum? maxSum:windowSum;
    }

    return maxSum;
}

int main() {
    vector<int> arr = {1,2,3,4,5};
    int k = 2;

    cout << maxSumSubarray(arr, k);

    return 0;
}